apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/AbstractSamhsaMatcherTest.java

```diff
@@ -0,0 +1,812 @@
+
+      doReturn(true).when(matcherSpy).isSamhsaIcdProcedure(mockConcept);
+
+      assertTrue(
+          "Samhsa procedure concept not correctly filtered",
+          matcherSpy.isSamhsaIcdProcedure(mockComponent));
+    }
+
+    @Test
+    public void shouldReturnFalseForIsSamhsaIcdProcedure() {
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      AbstractSamhsaMatcher<IBaseResource> matcherSpy = spy(AbstractSamhsaMatcher.class);
+
+      ProcedureComponent mockComponent = mock(ProcedureComponent.class);
+      CodeableConcept mockConcept = mock(CodeableConcept.class);
+
+      doReturn(mockConcept).when(mockComponent).getProcedureCodeableConcept();
+
+      doThrow(INVOCATION_EXCEPTION)
+          .when(matcherSpy)
+          .isSamhsaIcdProcedure(any(CodeableConcept.class));
+
+      doReturn(false).when(matcherSpy).isSamhsaIcdProcedure(mockConcept);
+
+      assertFalse(
+          "Samhsa procedure concept not correctly filtered",
+          matcherSpy.isSamhsaIcdProcedure(mockComponent));
+    }
+
+    @Test
+    public void shouldCorrectlyInvokeIsSamhsaCodingForDiagnosis() {
+      CodeableConcept mockConcept = mock(CodeableConcept.class);
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      AbstractSamhsaMatcher<IBaseResource> matcherSpy = spy(AbstractSamhsaMatcher.class);
+
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      doReturn(false)
+          .when(matcherSpy)
+          .isSamhsaCoding(any(CodeableConcept.class), any(Predicate.class), any(Predicate.class));
+
+      doReturn(false).when(matcherSpy).isSamhsaIcd9Diagnosis(any(Coding.class));
+
+      doReturn(false).when(matcherSpy).isSamhsaIcd10Diagnosis(any(Coding.class));
+
+      ArgumentCaptor<CodeableConcept> conceptCaptor =
+          ArgumentCaptor.forClass(CodeableConcept.class);
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      ArgumentCaptor<Predicate<Coding>> icd9CallCaptor = ArgumentCaptor.forClass(Predicate.class);
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      ArgumentCaptor<Predicate<Coding>> icd10CallCaptor = ArgumentCaptor.forClass(Predicate.class);
+
+      matcherSpy.isSamhsaDiagnosis(mockConcept);
+
+      // Capture the arguments used when it was invoked
+      verify(matcherSpy)
+          .isSamhsaCoding(
+              conceptCaptor.capture(), icd9CallCaptor.capture(), icd10CallCaptor.capture());
+
+      assertEquals(mockConcept, conceptCaptor.getValue());
+
+      Coding mockCoding = mock(Coding.class);
+
+      // Invoke the captured lambdas to check they were the right ones
+      icd9CallCaptor.getValue().test(mockCoding);
+      icd10CallCaptor.getValue().test(mockCoding);
+
+      verify(matcherSpy, times(1)).isSamhsaIcd9Diagnosis(mockCoding);
+      verify(matcherSpy, times(1)).isSamhsaIcd10Diagnosis(mockCoding);
```

Comment:
Hmm, in this case I'm mocking the ICD9 and ICD10 checks, so actually it's not testing any other methods.  This would only be checking the logic within `isSamhsaDiagnosis()`.  The later verifications are only to check that the methods within `isSamhsaDiagnosis()` were invoked, but they are not actually invoking any code.  Am I misunderstanding?

Reply:
The part where you are checking method invocations within `isSamhsaDiagnosis` is the part I'm talking about. Ideally, you want to test a method using its input and ensure the output is as expected; we shouldnt _care_ that a specific method was called inside the method, only that the output is as expected, otherwise the test is too tightly coupled to the implementation. 

It might be a symptom of the way the test was set up. Since the ICD9/10 checks are within the test class, I don't see why they should be mocked in the first place. I'm assuming we set up a spy for the class under test because its an abstract class and we cant instantiate a real instance, but then causes some wonkiness with having to overmock all the parts. In the past for testing abstract classes, I've made a test class that is basically an empty "test" implementation of the class that just accepts all the default implementations, so you can test the default abstract methods without having to mock every part of the class. Then you could just call `isSamhsaDiagnosis()` without needing to worry about the internals of what that method does, and just pass values in and ensure the proper output comes out. (The parameterized tests would benefit from this as well)

Reply:
I think this comes back to needing some team discussion around testing naming, breakdowns, and standards; can revisit this and other tests and perhaps categorize/name them on their granularity once we settle on some team testing standards.